File: Rust/heapsort.rs
pub fn heap_sort<T: Ord>(arr: &mut [T]) {
    if arr.len() <= 1 {
        return; // already sorted
    }

    heapify(arr);

    for end in (1..arr.len()).rev() {
        arr.swap(0, end);
        move_down(&mut arr[..end], 0);
    }
}


fn heapify<T: Ord>(arr: &mut [T]) {
    let last_parent = (arr.len() - 2) / 2;
    for i in (0..=last_parent).rev() {
        move_down(arr, i);
    }
}


fn move_down<T: Ord>(arr: &mut [T], mut root: usize) {
    let last = arr.len() - 1;
    loop {
        let left = 2 * root + 1;
        if left > last {
            break;
        }
        let right = left + 1;
        let max = if right <= last && arr[right] > arr[left] {
            right
        } else {
            left
        };

        if arr[max] > arr[root] {
            arr.swap(root, max);
        }
        root = max;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty() {
        let mut arr: Vec<i32> = Vec::new();
        heap_sort(&mut arr);
        assert_eq!(&arr, &[]);
    }

    #[test]
    fn single_element() {
        let mut arr = vec![1];
        heap_sort(&mut arr);
        assert_eq!(&arr, &[1]);
    }

    #[test]
    fn sorted_array() {
        let mut arr = vec![1, 2, 3, 4];
        heap_sort(&mut arr);
        assert_eq!(&arr, &[1, 2, 3, 4]);
    }

    #[test]
    fn unsorted_array() {
        let mut arr = vec![3, 4, 2, 1];
        heap_sort(&mut arr);
        assert_eq!(&arr, &[1, 2, 3, 4]);
    }

    #[test]
    fn odd_number_of_elements() {
        let mut arr = vec![3, 4, 2, 1, 7];
        heap_sort(&mut arr);
        assert_eq!(&arr, &[1, 2, 3, 4, 7]);
    }

    #[test]
    fn repeated_elements() {
        let mut arr = vec![542, 542, 542, 542];
        heap_sort(&mut arr);
        assert_eq!(&arr, &vec![542, 542, 542, 542]);
    }
}

File: Python/heapsort.py
# Heap Sort Algorithm in Python

def heapify(arr, i, length):
    left = 2 * i + 1
    right = 2 * i + 2
    max_index = i

    if left < length and arr[left] > arr[max_index]:
        max_index = left

    if right < length and arr[right] > arr[max_index]:
        max_index = right

    if max_index != i:
        swap(arr, max_index, i)
        heapify(arr, max_index, length)


def swap(arr, a, b):
    arr[a], arr[b] = arr[b], arr[a]


# sorting is in the ascending order
def heap_sort(arr):
    for i in range(math.floor(len(arr) / 2), -1, -1):
        heapify(arr, i, len(arr))

    length = len(arr)
    # reorganise the heap by removing the root
    for i in range(len(arr) - 1, 0, -1):
        swap(arr, 0, i)
        length -= 1
        heapify(arr, 0, length)
    print(arr)


arr = [76, 22, 21, 96, 70, 20]
heap_sort(arr)

# Output:
# [20, 21, 22, 70, 76, 96]
